### PR TITLE
Arena hard elad2

### PR DESCRIPTION
**Pull request by eladven**
**LLMaJ:**
1. Added examples for llm as judge using Arena Hard (`Evaluate your model on the Arena Hard benchmark using a custom LLMaJ` and  `Evaluate a judge model performance judging the Arena Hard Benchmark`)
2. Added `tasks.response_assessment.pairwise_comparative_rating.single_turn` to the possible tasks llama can run
Datasets added:
3. LlmaJ now pass the original dataset privacy_policy to the new generated dataset
4. Added  `WeightedWinRateCorrelation` metric for `tasks.response_assessment.pairwise_comparative_rating.single_turn` 

**Datasets additions:**
1. Add ArenaHard dataset for evaluation and meta-evaluation of llmaj
2. Reward bench (a dataset for meta evaluation)

**Templates:**
1. Arena Hard template + processors
2. ArenaHard-like Prometeous template 
3. Added class PairwiseComparativeRatingTemplate

**Small Changes:**
1. Added  DuplicateSplit Operator 
2. 4. Instance Metadata now contains "data_classification_policy"

**Consistency Breaking dataset changes**
1. Added shuffle to the universal_ner card as the dataset comes sorted by classes from HF

**Small fixes:**
1. Fixed calpnq card by adding test to unanswerable responses
2. fixed in HF Space loader - _map_wildcard_path_to_full_paths didn't use the revision
4. SelectFields operator now always keep the fields `["data_classification_policy", "recipe_metadata"]`
5. Join Operator won't delete  `["data_classification_policy", "recipe_metadata"]`
6. Bug fix in shuffle operation in PairwiseChoiceTemplate

**Left TODOs:** (open issues)
1. Support using both positions in llmaj
2. Add a metric for judge bias to positions in llmaj
7. Maybe also add a metric to the meta-eval in llmaj

### CODE REVIEW

File: .github/workflows/catalog_consistency.yml [hunk 1]
```diff
@@ -25,9 +25,12 @@ jobs:
      - uses: actions/setup-python@v5
        with:
          python-version: '3.9'
-         
-     - run: curl -LsSf https://astral.sh/uv/install.sh | sh
-     - run: uv pip install --system -e ".[tests]"
+         cache: 'pip' # caching pip dependencies
```
Comment (yoavkatz) [inline .github/workflows/catalog_consistency.yml:28]: This should be returned, as it was a change done by Elron to us uv instead of pip. Otherwise, we should consult @elronbandel .

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -7,7 +7,7 @@ repos:
     hooks:
       # Run the linter.
       - id: ruff
-        args: [ --fix ]
+        args: [ --fix,  --ignore, C901]
```
Comment (yoavkatz) [inline .pre-commit-config.yaml:10]: Why ignore all C901 error (too complex code).  We can ignore specific files if needed.

File: docs/docs/examples.rst [hunk 3]
```diff
@@ -89,3 +89,24 @@ The example adds two LLM judges, one that uses the ground truth (references) fro
...
+Demonstrates how to evaluate a user model on the Arena Hard benchmark, using an LLMaJ other than the GPT4.
+
+`Example code <https://github.com/IBM/unitxt/blob/main/examples/evaluate_a_model_using_arena_hard.py>`_
+
+Related documentation: :ref:`Evaluate a Model on Arena Hard Benchmark <arena_hard_evaluation>`.
+
+Evaluate a judge model performance judging the Arena Hard Benchmark
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+DDemonstrates how to evaluate the capabilities of a user model, to act as a judge on the Arena Hard benchmark.
```
Comment (yoavkatz) [inline docs/docs/examples.rst:104]:
```suggestion
Demonstrates how to evaluate the capabilities of a user model, to act as a judge on the Arena Hard benchmark.
```

File: docs/docs/examples.rst [hunk 4]
```diff
@@ -89,3 +89,24 @@ The example adds two LLM judges, one that uses the ground truth (references) fro
...
+The model is evaluated on it's capabilities to give a judgment that is in correlation with GPT4 judgment on the benchmark.
```
Comment (yoavkatz) [inline docs/docs/examples.rst:105]:
```suggestion
The model is evaluated on its capability to give a judgment that is in correlation with GPT4 judgment on the benchmark.
```

File: examples/evaluate_a_judge_model_capabilities_on_arena_hard.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+from unitxt.inference import (
+    HFPipelineBasedInferenceEngine,
+)
+from unitxt.standard import StandardRecipe
+
+model_id = "meta-llama/llama-3-70b-instruct"
+model_format = "formats.llama3_instruct"
+
+dataset = (
+    StandardRecipe(
```
Comment (yoavkatz) [inline examples/evaluate_a_judge_model_capabilities_on_arena_hard.py:11]: We should use load_dataset() and not StandardRecipe , which is an internal implementation.

Comment (yoavkatz) [reply examples/evaluate_a_judge_model_capabilities_on_arena_hard.py:11]: load_dataset(card="cards.arena_hard.response_assessment.pairwise_comparative_rating.both_games_gpt_4_judge, template="templates.response_assessment.pairwise_comparative_rating.arena_hard_with_shuffling".format=model_format,loader_limit=4)

File: examples/evaluate_a_judge_model_capabilities_on_arena_hard.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+# inference_model = OpenAiInferenceEngine(model_name=model, parameters=params)
+
+# Using Huggingface model
+inference_model = HFPipelineBasedInferenceEngine(
+    model_name=model_id, max_new_tokens=1024
+)
+
+predictions = inference_model.infer(dataset)
+scores = evaluate(predictions=predictions, data=dataset)
+# [print(item) for item in scores[0]["score"]["global"].items()]
```
Comment (yoavkatz) [inline examples/evaluate_a_judge_model_capabilities_on_arena_hard.py:37]: Should print the score using print_dict.

File: prepare/metrics/llm_as_judge/pairwise_rating/llama_3_ibm_genai_arena_hard_template.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    "templates.response_assessment.pairwise_comparative_rating.arena_hard",
+    "templates.response_assessment.pairwise_comparative_rating.arena_hard_with_shuffling",
+]
+for template in templates:
+    task = "pairwise_comparative_rating.single_turn"
+
+    gen_params = IbmGenAiInferenceEngineParams(max_new_tokens=2048, random_seed=42)
+    for model_id in model_list:
+        inference_model = IbmGenAiInferenceEngine(
+            model_name=model_id, parameters=gen_params
```
Comment (yoavkatz) [inline prepare/metrics/llm_as_judge/pairwise_rating/llama_3_ibm_genai_arena_hard_template.py:20]: We noew have a new way to passing params (just by passing max_new_tokens=2048, random_seed=42).  No needed for parameter object. The old way is still supported, but is deprecated. We should change all internal code to use it.

File: prepare/metrics/rag_answer_correctness.py [hunk 8]
```diff
@@ -49,117 +49,128 @@ def test_answer_correctness(task_data, catalog_name, global_target, instance_tar
...
+            "ground_truths": ["B C", "C"],
+            "answer": "B",
+        },
+        {  # recall is 1/3
+            "ground_truths": ["D E F"],
+            "answer": "B C D",
+        },
+    ]
 
-test_answer_correctness(
```
Comment (yoavkatz) [inline prepare/metrics/rag_answer_correctness.py:95]: How is this change related?

File: prepare/tasks/response_assessment/pairwise_comparative_rating/single_turn.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
+from unitxt.blocks import Task
+from unitxt.catalog import add_to_catalog
+
+add_to_catalog(
+    Task(
+        inputs={
```
Comment (yoavkatz) [inline prepare/tasks/response_assessment/pairwise_comparative_rating/single_turn.py:6]: In a recent PR, these are now called "input_fields", and "reference_fields" instead of "inputs" and "outputs". Again - we support for backward compatibility, but we should change all the internal code for this.

File: prepare/templates/rag/response_generation.py [hunk 10]
```diff
@@ -16,7 +16,7 @@
 
 add_to_catalog(
     MultiReferenceTemplate(
-        instruction="Answer the question, basing your answer on the context",
+        instruction="Answer the question, basing your answer on the context\n\n",
```
Comment (yoavkatz) [inline prepare/templates/rag/response_generation.py:19]: This is a non backward compatible change (in terms of results). I'm not sure why it's needed, it seems like more of an "format" issue them a template issue.  The format can place the instructions anywhere in the prompt.

File: src/unitxt/stream_operators.py [hunk 11]
```diff
@@ -102,6 +90,26 @@ def merge(self, multi_stream) -> List:
...
+
+        def assert_col_values_are_identical(
+            df: pd.DataFrame, col_name_1: str, col_name_2
+        ):
+            assert df.apply(
+                lambda row: str(row[col_name_1]) == str(row[col_name_2]),
+                axis=1,
+            ).all()
+
+        common_cols_to_verify = ["data_classification_policy", "recipe_metadata"]
```
Comment (yoavkatz) [inline src/unitxt/stream_operators.py:102]: Not sure I understand (also not documented), what happens when there are columns with same name in left and right?   Right now I think they are kept as separate values ,except "data_classification_policy" and "recipe_metadata" which must be the same.

File: utils/prepare_all_artifacts2.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import importlib.util
```
Comment (yoavkatz) [inline utils/prepare_all_artifacts2.py:1]: Is this code needed?

**Comment by yoavkatz** [comment]
I added my comments.    One additional point I see, is there is delicate code which does not have tests (e.g. calculating mean rate).

File: src/unitxt/stream_operators.py (hunk 11 above)
Comment (OfirArviv) [reply src/unitxt/stream_operators.py:102]: it will turn into "data_classification_policy_x" and "data_classification_policy_y" for example

File: prepare/metrics/rag_answer_correctness.py (hunk 8 above)
Comment (OfirArviv) [reply prepare/metrics/rag_answer_correctness.py:95]: could be some merge that got it

File: prepare/cards/universal_ner.py [hunk 13]
```diff
@@ -48,6 +51,7 @@
             requirements_list=["conllu"],
         ),
         preprocess_steps=[
+            Shuffle(page_size=sys.maxsize),
```
[the remark this answers is not in the document]
Comment (OfirArviv) [reply prepare/cards/universal_ner.py:54]: I have no idea why its in here

File: .github/workflows/catalog_consistency.yml (hunk 1 above)
Comment (OfirArviv) [reply .github/workflows/catalog_consistency.yml:28]: this is just a merge thing. It should be as in the master

File: .pre-commit-config.yaml (hunk 2 above)
Comment (OfirArviv) [reply .pre-commit-config.yaml]: done

File: .secrets.baseline [hunk 14]
```diff
@@ -3,7 +3,7 @@
     "files": "^.secrets.baseline$",
     "lines": null
   },
-  "generated_at": "2024-07-09T07:07:12Z",
```
Comment (OfirArviv) [inline .secrets.baseline:6]: @eladven why the change? It wasn't in my changes

File: examples/evaluate_a_judge_model_capabilities_on_arena_hard.py (hunk 5 above)
Comment (OfirArviv) [reply examples/evaluate_a_judge_model_capabilities_on_arena_hard.py:11]: @yoavkatz we cannot use loader_limit=4, because there is a join operation in the card. We have to load at least 10k examples to not get an error

File: examples/evaluate_a_judge_model_capabilities_on_arena_hard.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+)
+# We are evaluating only on a small subset, in order for the example to finish quickly.
+# The dataset full size if around 40k examples. You should use around 1k-4k in your evaluations.
+
+# params = IbmGenAiInferenceEngineParams(max_new_tokens=1024, random_seed=42)
+# inference_model = IbmGenAiInferenceEngine(model_name=model_id, parameters=params)
+
+# Using OpenAi model
+# params = OpenAiInferenceEngineParams(max_new_tokens=1024)
+# inference_model = OpenAiInferenceEngine(model_name=model, parameters=params)
```
Comment (OfirArviv) [inline examples/evaluate_a_judge_model_capabilities_on_arena_hard.py:28]: we cannot use HF example with llama3

File: prepare/cards/universal_ner.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (OfirArviv) [reply prepare/cards/universal_ner.py:54]: @perlitz  added it. Probably for BlueBench

File: prepare/metrics/llm_as_judge/pairwise_rating/llama_3_ibm_genai_arena_hard_template.py (hunk 7 above)
Comment (OfirArviv) [reply prepare/metrics/llm_as_judge/pairwise_rating/llama_3_ibm_genai_arena_hard_template.py:20]: ok. Thou again, all these "smart" coding practice are totally unreadable.

File: prepare/cards/universal_ner.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (OfirArviv) [reply prepare/cards/universal_ner.py:54]: @yoavkatz the original dataset is sorted by classes

[the remark this answers is not in the document]
Comment (yoavkatz) [reply prepare/cards/universal_ner.py:55]: Ok.  This needs to be noted in the issue (so it is noted in the release notes)

File: prepare/metrics/llm_as_judge/pairwise_rating/llama_3_ibm_genai_arena_hard_template.py (hunk 7 above)
Comment (yoavkatz) [reply prepare/metrics/llm_as_judge/pairwise_rating/llama_3_ibm_genai_arena_hard_template.py]: The parameters  are still there in a class:


class IbmGenAiInferenceEngineParamsMixin(Artifact):
    beam_width: Optional[int] = None
    decoding_method: Optional[Literal["greedy", "sample"]] = None
    include_stop_sequence: Optional[bool] = None
    length_penalty: Any = None
    max_new_tokens: Optional[int] = None
    min_new_tokens: Optional[int] = None
    random_seed: Optional[int] = None
    repetition_penalty: Optional[float] = None
    return_options: Any = None
    stop_sequences: Optional[List[str]] = None
    temperature: Optional[float] = None
    time_limit: Optional[int] = None
    top_k: Optional[int] = None
    top_p: Optional[float] = None
    truncate_input_tokens: Optional[int] = None
    typical_p: Optional[float] = None
    
     It's just that they are inherited, and not contained as a separate object.

File: src/unitxt/stream_operators.py (hunk 11 above)
Comment (OfirArviv) [reply src/unitxt/stream_operators.py:106]: added a comment explaining it

File: utils/prepare_all_artifacts2.py (hunk 12 above)
Comment (eladven) [reply utils/prepare_all_artifacts2.py:1]: I added this script by mistake. We should remove it

File: prepare/templates/rag/response_generation.py (hunk 10 above)
Comment (perlitz) [reply prepare/templates/rag/response_generation.py:19]:
```suggestion
        instruction="Answer the question, basing your answer on the context",
```

File: prepare/templates/rag/response_generation.py [hunk 16]
```diff
@@ -27,7 +27,7 @@
...
```
Comment (perlitz) [inline prepare/templates/rag/response_generation.py:30]:
```suggestion
        instruction="Answer the question, basing your answer on the context",
```

File: prepare/cards/universal_ner.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (perlitz) [reply prepare/cards/universal_ner.py:55]: sure

File: prepare/metrics/rag_answer_correctness.py (hunk 8 above)
Comment (OfirArviv) [reply prepare/metrics/rag_answer_correctness.py:95]: Yotam did it when working on rag

File: examples/evaluate_a_judge_model_capabilities_on_arena_hard.py [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+"""
+We are evaluating only on a small subset (by using "select(range(4)), in order for the example to finish quickly.
+The dataset full size if around 40k examples. You should use around 1k-4k in your evaluations.
+"""
+dataset = load_dataset(
+    f"card=cards.arena_hard.response_assessment.pairwise_comparative_rating.both_games_gpt_4_judge,"
+    "template=templates.response_assessment.pairwise_comparative_rating.arena_hard_with_shuffling,"
+    f"format={model_format}"
+)["test"].select(range(4))
```
Comment (yoavkatz) [inline examples/evaluate_a_judge_model_capabilities_on_arena_hard.py:16]: I think it's better not to use strings , but instead :

taset = load_dataset(
    card="cards.arena_hard.response_assessment.pairwise_comparative_rating.both_games_gpt_4_judge",
    template="templates.response_assessment.pairwise_comparative_rating.arena_hard_with_shuffling"
    format=model_format,
    max_test_instances=4
)["test"]

File: tests/library/test_recipe.py [hunk 18]
```diff
@@ -176,7 +176,7 @@ def test_standard_recipe_production_with_demos(self):
             ' "choices": ["yes", "not", "maybe"],'
             ' "answer": "maybe",'
             ' "options": [" A", " B", " C"],'
-            ' "metadata": {"template": "templates.qa.multiple_choice.with_topic.lm_eval_harness"}'
```
Comment (yoavkatz) [inline tests/library/test_recipe.py:179]: Notice the data_classificaiton policy now appears in two places.  In the metadata and in a dedicated field (see line 183).

File: prepare/metrics/rag_answer_correctness.py (hunk 8 above)
Comment (yoavkatz) [reply prepare/metrics/rag_answer_correctness.py:95]: I saw another PR onthese metric yesterday.  Can you review with  @perlitz - that these change is correct.

I want to make sure we don't revert / make unintended changes.

File: examples/evaluate_a_judge_model_capabilities_on_arena_hard.py (hunk 17 above)
Comment (OfirArviv) [reply examples/evaluate_a_judge_model_capabilities_on_arena_hard.py:16]: Oh ok. Didn't know it's possible (the interface doesn't tell you that :X).
btw notice that we cannot use max_test_instances=4
We have to load all the dataset because of the join operation

**Review by OfirArviv — APPROVED** [review]
